plugins/spectate/pause.cpp

```diff
@@ -0,0 +1,160 @@
+#include "pause.h"
+#include <VTableInterpose.h>
+#include <df/global_objects.h>
+#include <df/viewscreen_dwarfmodest.h>
+#include <df/announcements.h>
+#include <df/d_init.h>
+#include <df/ui.h>
+
+#include <algorithm>
+
+using namespace DFHack;
+using namespace Pausing;
+using namespace df::enums;
+
+namespace pausing {
+    AnnouncementLock announcementLock("monitor");
+    PlayerLock playerLock("monitor");
+
+    const size_t announcement_flag_arr_size = sizeof(decltype(df::announcements::flags)) / sizeof(df::announcement_flags);
+    bool state_saved = false; // indicates whether a restore state is ok
+    bool announcements_disabled = false; // indicates whether disable or restore was last enacted, could use a better name
+    bool saved_states[announcement_flag_arr_size]; // state to restore
+    bool locked_states[announcement_flag_arr_size]; // locked state (re-applied each frame)
+    bool allow_player_pause = true; // toggles player pause ability
+
+    using df::global::ui;
+    using namespace df::enums;
+    struct player_pause_hook : df::viewscreen_dwarfmodest {
+        typedef df::viewscreen_dwarfmodest interpose_base;
+        DEFINE_VMETHOD_INTERPOSE(void, feed, (std::set<df::interface_key>* input)) {
+            if ((ui->main.mode == ui_sidebar_mode::Default) && !allow_player_pause) {
+                input->erase(interface_key::D_PAUSE);
+            }
+            INTERPOSE_NEXT(feed)(input);
+        }
+    };
+
+    IMPLEMENT_VMETHOD_INTERPOSE(player_pause_hook, feed);
+}
+using namespace pausing;
+
+std::unordered_set<Lock*> PlayerLock::locks;
+std::unordered_set<Lock*> AnnouncementLock::locks;
```

Comment:
From [Discord](https://discord.com/channels/793331351645323264/857489615299411968/1030355598556012555): these sets appear to be initialized after the `announcementLock` and `playerLock` above are initialized, which causes undefined behavior, and is likely the cause of the tests hanging.

Moving these definitions above `namespace pausing` seems to help, but I'm not sure if that is truly sufficient.

Reply:
Should be, but i'm no expert in the written law of this land. It works, but afaik something like this could be UB I know static initialization can get a bit complainy. I'd expect a warning [error] if it were. Would enabling sanitiziers on debug builds be feasible?